src/sql/parts/registeredServer/common/objectExplorerService.ts

```diff
@@ -423,4 +445,188 @@ export class ObjectExplorerService implements IObjectExplorerService {
 	public getServerTreeView() {
 		return this._serverTreeView;
 	}
+
+	public getActiveConnections(): { connectionId: string, nodeInfo: sqlops.NodeInfo }[] {
+		let connections: [string, TreeNode][] = Object.entries(this._activeObjectExplorerNodes);
+		return connections.map(([connectionId, treeNode]) => {
+			return {
+				connectionId: connectionId,
+				nodeInfo: treeNode.toNodeInfo()
+			};
+		});
+	}
+
+	public expandNodeForConnection(connectionId: string, nodePath: string, stopAtParent: boolean = false): Thenable<void> {
+		let rootNode = this._activeObjectExplorerNodes[connectionId];
+		if (!rootNode) {
+			return Promise.reject('The given connection is not active in Object Explorer');
+		}
+		let node = this.findTreeNode(connectionId, nodePath);
+		if (!node) {
+			// The node is not in the tree yet, so expand its parent nodes to make it visible
+			return this.findNodeInfo(connectionId, nodePath).then(nodeInfo => {
```

Comment:
Minor: I've been using async/await recently and it simplifies things a lot, especially where we have nested thenables with promises, more thenables, etc. Are we using it inside our ops studio code yet? This function seems like a nice place to try it out.

Reply:
Also given the logic here, maybe just wrap this in a private method so you get a simpler flow. There's a lot of steps in each sub-clause, usually means we should separate into their own methods.
```
if (!node) {
    return this.findAndSetExpanded(...);
} else { 
    return this.expandToNode(...);
}
```

Reply:
Thanks for the async/await tip, it made this code much cleaner!

Reply:
I would recommend not using async await, although it might make the code look pretty; because we are targeting an older version of javascript, it makes debugging nearly impossible. Just take a look at the output of a async await function.

Reply:
@anthonydresser can you explain what you mean? I did a lot of debugging while trying to get this code right and didn't run into any major problems, but maybe I just got lucky. In fact I almost felt that it was easier to debug since now I can just step over async calls if I want to

Reply:
I found that when I tried to step into and through to await functions I would often be facing the underlying generator/await that is the javascript generated from using async/await, which forces you to debug very ugly javascript. Also, you have to be careful when it comes to errors, to properly get errors from await functions you have to wrap them in try/catches.